### PR TITLE
Pass the index into the user defined filter function

### PR DESCRIPTION
**Pull request by filaraujo**
This patch passes the index into the user defined filter function, this allows the user to define any function and receive the similiar signature that they would expect in a Array.filter call

### CODE REVIEW

**Comment by filaraujo** [comment]
how can i rerun these tests?

File: src/lib/template/dom-repeat.html [hunk 1]
```diff
@@ -370,7 +370,7 @@
       // Apply user filter to keys
       if (this._filterFn) {
         this._orderedKeys = this._orderedKeys.filter(function(a) {
-          return this._filterFn(c.getItem(a));
```
Comment (samccone) [inline src/lib/template/dom-repeat.html]: to be totally inline with the filter api we should also pass the original array as well, thoughts?

**Comment by ebidel** [comment]
@filaraujo Can you rebase and add the other `Array.filter` args? We should be 1:1 with the native api.

**Comment by filaraujo** [comment]
Looking at the latest code, index and items would be troublesome and misleading.  

Unsure which index to pass. Do I pass the virtual index ( the one that references the internal collection) or do I pass the current index of the filter loop.  Passing the filter iteration loop doesn't really provide much use.  

As for the items, I can pass the original collection or the references to the weak map collection.  

Thoughts?

**Comment by kevinpschaaf** [comment]
Due to the API inconsistency that would result due to differences in when we can efficiently pass the index and when it would be prohibitive to do so (because the array is operating on sorted keys, see https://github.com/Polymer/polymer/issues/2187#issuecomment-127452542), we're electing to not merge this now.  Thanks for the efforts here, this is hard problem.

**Pull request closed without merging**